File: 300-longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1] * n
        ans = 1  # Initialize ans to 1 since the minimum LIS is always 1

        for i in range(1, n):
            maxx = 0
            for j in range(0, i):
                if nums[j] < nums[i]:
                    maxx = max(maxx, dp[j])

            dp[i] = maxx + 1
            ans = max(ans, dp[i])

        return ans

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
import java.util.Arrays;

class Solution {
    public int lengthOfLIS(int[] nums) {
        if (nums == null || nums.length == 0) return 0;
        int n = nums.length;
        int[] dp = new int[n];
        Arrays.fill(dp, 1); 
        int maxLIS = 1;

        for (int i = 1; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i]) {
                    int newLen=dp[j]+1;
                    dp[i]=Math.max(dp[i],newLen);
                    // dp[i] = Math.max(dp[i], dp[j] + 1); 
                }
            }
            maxLIS = Math.max(maxLIS, dp[i]); 
        }

        return maxLIS;
    }
}
